Formatting

dpsutton opened the pull request:
return a format string so it can be used over multiple rows. The rules
for formatting are quite annoying because there are defaults that are
assumed rather than present. IE, if you set a column as currency it
doesn't set the number style sometimes and defaults to USD without that
being in the settings. In the future would be far preferable to always
create a fully fleshed map of defaults rather than inferring them in
random places.

dpsutton committed:
Formatting

return a format string so it can be used over multiple rows. The rules
for formatting are quite annoying because there are defaults that are
assumed rather than present. IE, if you set a column as currency it
doesn't set the number style sometimes and defaults to USD without that
being in the settings. In the future would be far preferable to always
create a fully fleshed map of defaults rather than inferring them in
random places.

dpsutton committed:
appease our doc string supervisor

dpsutton committed:
Include column information for formatting numbers

really important for requiring two decimal places by default on doubles
and none on integers. Its close to correct ignoring it and using
"#,###.##" and this leaves non decimal places on integers and allows
only up to two on doubles, but things like 31.10 will format as 31.1 and
that is improper

dpsutton committed:
Remove unneeded metabase.test

dpsutton committed:
Update tests with nil column

noahmoss commented on src/metabase/pulse/render/common.clj:
```diff
@@ -22,11 +26,61 @@
+       (if (number? value)
+         (NumericWrapper.
+          (str (when prefix prefix)
+               (when (and currency? (or (nil? currency-style)
+                                        (= currency-style "symbol")))
+                 (get-in currency/currency [(keyword (or currency "USD")) :symbol]))
+               (when (and currency? (= currency-style "code"))
+                 (str (get-in currency/currency [(keyword (or currency "USD")) :code]) \space))
+               (.format fmtr (* value (or scale 1)))
+               (when (and currency? (= currency-style "name"))
+                 (str \space (get-in currency/currency [(keyword (or currency "USD")) :name_plural])))
```
On the frontend, the key `mv.viz/currency-in-header` key determines whether a currency value in a table should be formatted -- it should only be explicitly formatted when it is `false`, not when it's `true` or omitted.

Unfortunately it seems that in chart axes, currency values are always formatted regardless of this key. So I'm not sure if there's a way to get this right unless this code has context about whether a value will be rendered in a table or a chart. 😕

noahmoss reviewed (approved):
LGTM, checked it out locally and tested some (non-exhaustive) edge cases. Just a few non-blocking comments & nits.

It would be really nice to be able to reuse this stuff for JSON & CSV formatting, as well as any future formatting use cases that might arise in the future. That would probably require extracting the logic into a shared library & adding the ability to granularity enable/disable certain options depending on the limitations of the target format.

dpsutton commented on src/metabase/pulse/render/body.clj:
```diff
@@ -122,31 +143,41 @@
+                         ;; todo remapped cols
+                         (map (fn [[col viz]]
+                                (get-format timezone-id col viz)))
+                         (for [{id :id col-name :name :as col} cols]
```
this is the correct formatting level. these two are arguments to `into`

dpsutton commented on src/metabase/pulse/render/common.clj:
```diff
@@ -22,11 +26,61 @@
+  ([column col-viz]
+   (number-formatter column
+                     col-viz
+                     (mb.viz/db->norm-column-settings-entries
+                      (:type/Number (public-settings/custom-formatting)))))
```
Yeah. I read over the xlsx stuff and I was worried that combining the two would make everything harder to read and not work as well. The xlsx stuff only returns format strings and then sets them on some reused cell objects and the requirements to unify it felt like everything would be harder to understand with lots of stuff passed through to keep track if we wanted an xlsx format string, a value formatting string, etc.

I think you might be right about the merge-global-settings though.

dpsutton replied:
Actually, I think there might be another way to approach this. I suspect we don't want to read this setting a bunch of times. It might be better to have the `process-viz-settings?` middleware to grab the settings once rather than repeatedly in other places. then a simple helper function can get the format settings for a column by getting by column id or name and merging the global settings.

dpsutton committed:
Get global viz settings in middleware, pass entire to formatting

Passing the entire viz settings to the formatting stuff makes the
calling code much simpler and the places where we merge global settings,
possibly get currency settings, etc, just in one place. We already had
the col information there so just use it.

dpsutton committed:
Remove public settings now that we get it from the metadata
